File: clients/typescript/src/accounts.ts
import type { Transport } from "./transport.js";

/** The active account scope returned by `POST /api/accounts/:id/activate`. */
export interface AccountScope {
  account: string;
  role: string;
}

/** Multi-tenancy account operations (requires ZigBase >= 0.9.0 with `.tenancy` enabled). */
export class AccountsService {
  constructor(private readonly transport: Transport) {}

  /**
   * POST /api/accounts/:id/activate — verifies an ACTIVE membership, sets the signed
   * `zb_account` cookie (same-origin browser apps; the default `credentials: "same-origin"`
   * keeps it), and returns the scope. 403 when not a member; 404 when tenancy is disabled.
   * API/SSR clients should prefer `client.withAccount(id)` / the `accountId` option —
   * the SDK never reads the cookie.
   */
  activate(
    accountId: string,
    opts: { signal?: AbortSignal; requestKey?: string } = {},
  ): Promise<AccountScope> {
    return this.transport.send<AccountScope>(
      `/api/accounts/${encodeURIComponent(accountId)}/activate`,
      { method: "POST", signal: opts.signal, requestKey: opts.requestKey },
    );
  }
}

File: clients/dart/lib/src/accounts.dart
/// Multi-tenancy account activation (requires ZigBase >= 0.9.0 with
/// `.tenancy` enabled).
///
/// A port of `AccountsService` in `clients/typescript/src/accounts.ts`; the
/// wire envelope (`{account, role}`) is confirmed against `src/api/accounts.zig`.
library;

import 'transport.dart';

/// The active account scope returned by `POST /api/accounts/:id/activate`.
class AccountScope {
  final String account;
  final String role;

  AccountScope({required this.account, required this.role});

  factory AccountScope.fromJson(Map<String, dynamic> json) {
    return AccountScope(
      account: json['account'] as String,
      role: json['role'] as String,
    );
  }
}

/// Multi-tenancy account operations.
class AccountsService {
  final Transport _transport;

  AccountsService(this._transport);

  /// `POST /api/accounts/:id/activate` — verifies an ACTIVE membership, sets
  /// the signed `zb_account` cookie (same-origin browser apps), and returns
  /// the scope. 403 when not a member; 404 when tenancy is disabled. API/SSR
  /// clients should prefer a dedicated `X-Account-Id`-scoped client — the SDK
  /// never reads the cookie itself.
  Future<AccountScope> activate(String accountId) async {
    final res = await _transport.send(
      '/api/accounts/${Uri.encodeComponent(accountId)}/activate',
      method: 'POST',
    ) as Map<String, dynamic>;
    return AccountScope.fromJson(res);
  }
}
